### PR TITLE
[ROCM] enabling gemm_rewriter subtests and several HLO passes for AMDGPU compiler

### PR DESCRIPTION
**Pull request by pemeliya**
We enable the majority of gemm_rewriter subtests on ROCM as well as activated DotDimensionMerger and PadForGemms 
HLO passes on ROCm.

The test xla/service/gpu/tests/gemm_rewrite_test.cc is very large. In order to easily enable/disable individual subtests for CUDA and ROCM paltforms, I have integrated the check function, which, in its most generic form, is implemented as follows:
```
  bool CudaOrRocmCheck(
          absl::AnyInvocable<bool(se::CudaComputeCapability)> cuda_fun,
          absl::AnyInvocable<bool(se::RocmComputeCapability)> rocm_fun) {
    const auto& gcomp = GpuComputeComp();            
    if(auto pcuda = std::get_if< se::CudaComputeCapability >(&gcomp); pcuda) {
      return (cuda_fun ? cuda_fun(*pcuda) : true); 
    } else {
      auto rocm = std::get< se::RocmComputeCapability >(gcomp);
      return (rocm_fun ? rocm_fun(rocm) : true); 
    }
```

Besides, I also override gtest ```void setUp()``` function to conditionally skip the whole test-suites based on the target architecture. The CUDA part is almost always not touched: only ```CudaCompileCapability``` is replaced with a generic variant ```GpuCompileCapability``` and respective checks.

@xla-rotation: could you please have a look ?

### CODE REVIEW

File: xla/service/gpu/amdgpu_compiler.cc
Comment (ezhulenev) [inline xla/service/gpu/amdgpu_compiler.cc:131]: remove?

File: xla/service/gpu/cublas_padding_requirements.cc
Comment (ezhulenev) [inline xla/service/gpu/cublas_padding_requirements.cc:32]: `auto* pcc` to make it clear that it's a pointer

even better use a std::visit here

```
struct RequiresPaddingVisitor {
  bool operator(se::CudaComputeCapability cc) {}
}

return std::visit(RequiresPaddingVisitor{}, cc);
```

File: xla/service/gpu/cublas_padding_requirements.h
Comment (ezhulenev) [inline xla/service/gpu/cublas_padding_requirements.h:35]: remove?

File: xla/service/gpu/gemm_rewriter_triton.cc
Comment (ezhulenev) [inline xla/service/gpu/gemm_rewriter_triton.cc:1282]: formatting is of, and again std visitor would be better here

**Comment by ezhulenev** [comment]
we dispatch to CUDA/ROCM CC many time here, I'd suggest to add overload pattern to gemm_rewriter and replace all if/else with an std::visit (https://www.modernescpp.com/index.php/visiting-a-std-variant-with-the-overload-pattern/)

File: xla/service/gpu/cublas_padding_requirements.cc
Comment (pemeliya) [reply xla/service/gpu/cublas_padding_requirements.cc]: yes, std::visit is a good idea, indeed !

**Review by ezhulenev — APPROVED** [review]

**Comment by ddunl** [comment]
I think needs rebase before it can be imported (+ CI fix)

**Review by akuegel — APPROVED** [review]

**Comment by pemeliya** [comment]
> I think needs rebase before it can be imported (+ CI fix)

fixed! there was a missing const reference in triton function after rebase

**Review by akuegel — APPROVED** [review]

**Review by akuegel — APPROVED** [review]

File: xla/service/gpu/tests/gemm_rewrite_test.cc
Comment (pemeliya) [reply xla/service/gpu/tests/gemm_rewrite_test.cc:23]: suprisingly, bazel did not complain about this missing dependency, but I have added it anyway

**Review by akuegel — APPROVED** [review]

**Comment by pemeliya** [comment]
I wonder, if there anything still needs to be done on this PR ? Thanks!

**Comment by akuegel** [comment]
> I wonder, if there anything still needs to be done on this PR ? Thanks!

Waiting for internal approval by someone else than me. In our process, the one who approves on GitHub cannot also approve internally.

@xla-rotation PTAL

**Pull request closed without merging**
